src/maptm/src/maptm_config.c

```diff
@@ -0,0 +1,526 @@
+    }
+    else if ((length >= 33) && (length <= 64))
+    {
+        mask = htonl((uint32_t)-1 << (64 - length));
+        if ((iapd.s6_addr32[0] == pref.s6_addr32[0])
+            && (iapd.s6_addr32[1] & mask) == (pref.s6_addr32[1] & mask))
+        {
+            return 1;
+        }
+    }
+    else if ((length >= 65) && (length <= 96))
+    {
+        mask = htonl((uint32_t)-1 << (96 - length));
+        if ((iapd.s6_addr32[0] == pref.s6_addr32[0])
+            && (iapd.s6_addr32[1] == pref.s6_addr32[1])
+            && (iapd.s6_addr32[2] & mask) == (pref.s6_addr32[2] & mask))
+        {
+            return 1;
+        }
+    }
+    else if ((length >= 97) && (length <= 128))
+    {
+        mask = htonl((uint32_t)-1 << (128 - length));
+        if ((iapd.s6_addr32[0] == pref.s6_addr32[0])
+            && (iapd.s6_addr32[1] == pref.s6_addr32[1])
+            && (iapd.s6_addr32[2] == pref.s6_addr32[2])
+            && (iapd.s6_addr32[3] & mask) == (pref.s6_addr32[3] & mask))
+        {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+// Select matched MAP-T rule
+struct mapt* get_Mapt_Rule(char *option95, char *iapd)
+{
+    LOGT("Get MAP-t Rules");
+    if (!option95) return NULL;
+
+    bool ret = false;
+    char *mapt_option95 = NULL;
+    mapt_option95 = option95;
+
+    ds_dlist_t l_rules;
+    ds_dlist_init(&l_rules, struct list_rules, d_node);
+
+    // Fill MAP-T rules list
+    char *rule = strtok(mapt_option95, " ");
+    while (rule != NULL)
+    {
+        struct list_rules *l_node;
+        l_node = malloc(sizeof(struct list_rules));
+        if (l_node == NULL)
+        {
+            LOGE("Unable to allocate update handler!");
+            free(l_node);
+            return NULL;
+        }
+
+        l_node->value = strdup(rule);
+        if ((l_node->value) == NULL)
+        {
+            LOGE("Unable to allocate update handler!");
+            free(l_node);
+            maptm_remove_list(l_rules);
+            return NULL;
+        }
+
+        ds_dlist_insert_tail(&l_rules, l_node);
+        rule = strtok(NULL, " ");
+    }
+
+    struct list_rules *node = NULL;
+    ds_dlist_iter_t iter;
+    struct mapt *mapt_rule = NULL;
+
+    for (   node = ds_dlist_ifirst(&iter, &l_rules);
+            node != NULL;
+            node = ds_dlist_inext(&iter))
+    {
+        mapt_rule = parse_option_rule(node->value);
+        if (mapt_rule)
+        {
+            ret = comparePrefix(iapd, mapt_rule->ipv6prefix, mapt_rule->prefix6len);
+            if (ret)
+            {
+                LOGD("MAP-T rule found");
+                break;
+            }
+            else
+            {
+                maptm_remove_maptStruct(mapt_rule);
+            }
+        }
+    }
+
+    if (ret)
+    {
+        maptm_remove_list(l_rules);
+
+        if (!ret)
+        {
+            LOGD("MAP-T rule NOT found");
+            return NULL;
+        }
+
+    return mapt_rule;
+        return mapt_rule;
+    }
+    return NULL;
```

Comment:
```suggestion
    maptm_remove_list(l_rules);

    if (!ret)
    {
        LOGD("MAP-T rule NOT found");
        return NULL;
    }

    return mapt_rule;
```

The code proposed in an earlier conversation (https://github.com/plume-design/opensync/pull/17/files#r461631386) was incorrectly applied.